File: app/src/main/java/com/testproject/kaera/ringtestapp/di/modules/ApiModule.java
package com.testproject.kaera.ringtestapp.di.modules;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;
import com.testproject.kaera.ringtestapp.enteties.APIRedditItem;
import com.testproject.kaera.ringtestapp.service.adapter.APIRedditItemAdapter;
import com.testproject.kaera.ringtestapp.util.curl.CurlInterceptor;

import java.lang.reflect.Type;
import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.techery.janet.http.HttpClient;
import io.techery.janet.okhttp3.OkClient;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;

@Module
public class ApiModule {

    @Provides
    @Singleton
    Gson provideGson() {
        Type apiRedditItemType = new TypeToken<List<APIRedditItem>>() {}.getType();
        return new GsonBuilder().serializeNulls()
                .registerTypeAdapter(apiRedditItemType, new APIRedditItemAdapter())
                .disableHtmlEscaping().create();
    }

    @Provides
    @Singleton
    OkHttpClient provideOkHttp() {
        HttpLoggingInterceptor logInterceptor = new HttpLoggingInterceptor();
        logInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .connectTimeout(20L, TimeUnit.SECONDS)
                .readTimeout(20L, TimeUnit.SECONDS)
                .addInterceptor(new CurlInterceptor())
                .addInterceptor(logInterceptor);
        return builder.build();
    }

    @Provides
    @Singleton
    HttpClient provideHttpClient(OkHttpClient okHttpClient) {
        return new OkClient(okHttpClient);
    }
}

File: app/src/main/java/com/testproject/kaera/ringtestapp/di/modules/ApiModule.kt
package com.testproject.kaera.ringtestapp.di.modules

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.testproject.kaera.ringtestapp.enteties.APIRedditItem
import com.testproject.kaera.ringtestapp.service.adapter.APIRedditItemAdapter
import com.testproject.kaera.ringtestapp.util.curl.CurlInterceptor
import dagger.Module
import dagger.Provides
import io.techery.janet.http.HttpClient
import io.techery.janet.okhttp3.OkClient
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Module
class ApiModule @Inject constructor() {

    @Provides
    @Singleton
    fun provideGson(): Gson {
        val apiRedditItemType = object : TypeToken<List<APIRedditItem>>() {

        }.type
        return GsonBuilder().serializeNulls()
                .registerTypeAdapter(apiRedditItemType, APIRedditItemAdapter())
                .disableHtmlEscaping().create()
    }

    @Provides
    @Singleton
    fun provideOkHttp(): OkHttpClient {
        val logInterceptor = HttpLoggingInterceptor()
        logInterceptor.level = HttpLoggingInterceptor.Level.BODY

        val builder = OkHttpClient.Builder()
                .connectTimeout(20L, TimeUnit.SECONDS)
                .readTimeout(20L, TimeUnit.SECONDS)
                .addInterceptor(CurlInterceptor())
                .addInterceptor(logInterceptor)
        return builder.build()
    }

    @Provides
    @Singleton
    fun provideHttpClient(okHttpClient: OkHttpClient): HttpClient {
        return OkClient(okHttpClient)
    }
}